File: LR1/src/main/scala/cycle_list/CycleList.scala
package cycle_list

import comparator.Comparator
import types.users.UserType
import java.io._

class CycleList extends Serializable {
  private var head: Node = _
  private var length: Int = 0

  private class Node(var data: AnyRef) extends Serializable {
    var next: Node = _
    var prev: Node = _

    override def toString: String = data.toString
  }

  def add(data: AnyRef): Unit = {
    if (head == null) {
      var node = new Node(data)
      node.next = node
      node.prev = node
      head = node
      length += 1
      return
    }
    var tail: Node = head.prev
    var node: Node = new Node(data)
    node.next = head
    head.prev = node
    node.prev = tail
    tail.next = node
    length += 1
  }

  def add(data: AnyRef, index: Int): Unit = {
    var tmp = getNode(index)
    var newNode = new Node(data)
    var tail = head.prev
    if (tmp ne head) {
      tmp.prev.next = newNode
      newNode.prev = tmp.prev
    }
    else head = newNode
    newNode.next = tmp
    tmp.prev = newNode
    tail.next = head
    head.prev = tail
    length += 1
  }

  def remove(index: Int): Unit = {
    var tmp = getNode(index)
    var tail: Node = head.prev
    if (tmp ne head) tmp.prev.next = tmp.next
    else head = tmp.next
    if (tmp ne tail) tmp.next.prev = tmp.prev
    else tail = tmp.prev
    tmp.next = null
    tmp.prev = null
    tail.next = head
    head.prev = tail
    length -= 1
  }

  def getByIndex(index: Int): AnyRef = getNode(index).data

  def getLength: Int = length

  private def getNode(index: Int): Node = {
    if (index < 0 || index >= length) throw new IndexOutOfBoundsException
    var tmp = head
    for (i <- 0 until index) {
      tmp = tmp.next
    }
    tmp
  }

  @throws[IOException]
  def forEach(iterator: Iterator): Unit = {
    var tmp = head
    for (i <- 0 until length) {
      iterator.toDo(tmp.data)
      tmp = tmp.next
    }
  }

  @throws[IOException]
  def forEachReverse(iterator: Iterator): Unit = {
    var tmp = head
    tmp = tmp.prev
    for (i <- 0 until length) {
      iterator.toDo(tmp.data)
      tmp = tmp.prev
    }
  }

  def sort(comparator: Comparator): Unit = {
    if (head != null && (head.next ne head) && (head.prev ne head)) {
      var tail = head.prev
      tail.next = null
      head.prev = null
      head = mergeSort(head, comparator)
      tail = getNode(length - 1)
      tail.next = head
      head.prev = tail
    }
  }

  private def mergeSort(headNode: Node, comparator: Comparator): Node = {
    if (headNode == null || headNode.next == null) return headNode
    val middle = getMidNode(headNode)
    val middleNext = middle.next
    middle.next = null
    val left = mergeSort(headNode, comparator)
    val right = mergeSort(middleNext, comparator)
    merge(left, right, comparator)
  }

  private def merge(first: Node, second: Node, comparator: Comparator): Node = {
    var firstNode: Node = first
    var secondNode: Node = second
    val merged = new Node(null)
    var temp = merged
    var tail = head.prev
    while (firstNode != null && secondNode != null) {
      if (comparator.compare(firstNode.data, secondNode.data) < 0) {
        temp.next = firstNode
        firstNode.prev = temp
        firstNode = firstNode.next
      }
      else {
        temp.next = secondNode
        secondNode.prev = temp
        secondNode = secondNode.next
      }
      temp = temp.next
    }
    while (firstNode != null) {
      temp.next = firstNode
      firstNode.prev = temp
      firstNode = firstNode.next
      temp = temp.next
    }
    while (secondNode != null) {
      temp.next = secondNode
      secondNode.prev = temp
      secondNode = secondNode.next
      temp = temp.next
      tail = temp
    }
    merged.next
  }

  private def getMidNode(node: Node): Node = {
    var previousNode = node
    var currentNode = node
    while (currentNode.next != null && currentNode.next.next != null) {
      previousNode = previousNode.next
      currentNode = currentNode.next.next
    }
    previousNode
  }

  def printList(): Unit = {
    var tmp = head
    for (i <- 0 until length) {
      System.out.print(i + ") ")
      System.out.println(tmp.data)
      tmp = tmp.next
    }
  }

  override def toString: String = {
    var str = ""
    var tmp = head
    for (i <- 0 until length) {
      str = str + tmp.data + "\n"
      tmp = tmp.next
    }
    str
  }

  def clearList(): Unit = {
    head = null
    length = 0
  }

  def save(userType: UserType, fileName: String): Unit = {
    try {
      val writer = new BufferedWriter(new FileWriter(fileName))
      try {
        writer.write(userType.typeName + "\n")
        this.forEach((data: AnyRef) => writer.write(data.toString + "\n"))
      } catch {
        case e: IOException =>
          e.printStackTrace()
      } finally if (writer != null) writer.close()
    }
  }

  def load(userType: UserType, fileName: String): Unit = {
    clearList()
    try {
      val br = new BufferedReader(new FileReader(fileName))
      try {
        var line: String = null
        line = br.readLine
        if (!(userType.typeName == line)) throw new Exception("Wrong file structure")
        line = br.readLine
        while (line != null) {
          add(userType.parseValue(line))
          line = br.readLine
        }
      } catch {
        case e: Exception =>
          throw new RuntimeException(e)
      } finally if (br != null) br.close()
    }
  }

  // func

  def mergeSortFuncStyle(comparator: Comparator): CycleList = {
    if (this.length <= 1)
      this
    else {
      val sortedList = new CycleList
      var leftList = new CycleList
      var rightList = new CycleList
      val middle = this.length / 2;
      forEachListLeft(x => {
        leftList.add(x)
      }, 0, middle)

      forEachListRight(x => {
        rightList.add(x)
      }, middle, this.length)

      leftList = leftList.mergeSortFuncStyle(comparator)
      rightList = rightList.mergeSortFuncStyle(comparator)

      var left = leftList.head
      var right = rightList.head

      while (leftList.length > 0 && rightList.length > 0) {
        if (comparator.compare(left.data, right.data) <= 0) {
          sortedList.add(left.data)
          left = left.next
          leftList.remove(0)
        } else {
          sortedList.add(right.data)
          right = right.next
          rightList.remove(0)
        }
      }

      while (leftList.length > 0) {
        sortedList.add(left.data)
        left = left.next
        leftList.remove(0)
      }

      while (rightList.length > 0) {
        sortedList.add(right.data)
        right = right.next
        rightList.remove(0)
      }

      sortedList
    }
  }

  def forEachListLeft(iterator: Iterator, idBegin: Int, idEnd: Int): Unit = {
    var tmp = head
    for (i <- idBegin until idEnd) {
      iterator.toDo(tmp.data)
      tmp = tmp.next
    }
  }

  def forEachListRight(iterator: Iterator, idBegin: Int, idEnd: Int): Unit = {
    var tmp = head
    for (i <- 0 until idBegin) {
      tmp = tmp.next
    }
    for (i <- idBegin until idEnd) {
      iterator.toDo(tmp.data)
      tmp = tmp.next
    }
  }

}

File: LR0/src/CycleList/CycleList.java
package CycleList;

import Comparator.Comparator;
import Types.Users.UserType;

import java.io.*;

public class CycleList implements Serializable {

    private Node head;
    private int length;

    private class Node implements Serializable {
        Object data;
        Node next;
        Node prev;

        public Node(Object data) {
            this.data = data;
        }

        @Override
        public String toString() {
            return data.toString();
        }
    }

    public void add(Object data) {
        if (head == null) {
            Node node = new Node(data);
            node.next = node.prev = node;
            head = node;
            length++;
            return;
        }
        Node tail = head.prev;
        Node node = new Node(data);
        node.next = head;
        head.prev = node;
        node.prev = tail;
        tail.next = node;
        length++;
    }

    public void add(Object data, int index) {
        Node tmp = getNode(index);
        Node newNode = new Node(data);
        Node tail = head.prev;
        if (tmp != head) {
            tmp.prev.next = newNode;
            newNode.prev = tmp.prev;
        } else {
            head = newNode;
        }
        newNode.next = tmp;
        tmp.prev = newNode;
        tail.next = head;
        head.prev = tail;
        length++;
    }

    public void remove(int index) {
        Node tmp = getNode(index);
        Node tail = head.prev;
        if (tmp != head) {
            tmp.prev.next = tmp.next;
        } else {
            head = tmp.next;
        }
        if (tmp != tail) {
            tmp.next.prev = tmp.prev;
        } else {
            tail = tmp.prev;
        }
        tmp.next = tmp.prev = null;
        tail.next = head;
        head.prev = tail;
        length--;
    }

    public Object getByIndex(int index) {
        return getNode(index).data;
    }

    public int getLength() {
        return length;
    }

    private Node getNode(int index) {
        if (index < 0 || index >= length) throw new IndexOutOfBoundsException();
        Node tmp = head;

        for (int i = 0; i < index; i++) {
            tmp = tmp.next;
        }
        return tmp;
    }

    public void forEach(Iterator iterator) throws IOException {
        Node tmp = head;
        for (int i = 0; i < length; i++) {
            iterator.toDo(tmp.data);
            tmp = tmp.next;
        }
    }

    public void forEachReverse(Iterator iterator) throws IOException {
        Node tmp = head;
        tmp = tmp.prev;
        for (int i = 0; i < length; i++) {
            iterator.toDo(tmp.data);
            tmp = tmp.prev;
        }
    }


    public void sort(Comparator comparator) {
        if (head != null && head.next != head && head.prev != head) {
            Node tail = head.prev;
            tail.next = null;
            head.prev = null;
            head = mergeSort(head, comparator);
            tail = getNode(length - 1);
            tail.next = head;
            head.prev = tail;
        }
    }

    private Node mergeSort(Node headNode, Comparator comparator) {
        if (headNode == null || headNode.next == null) {
            return headNode;
        }
        Node middle = getMidNode(headNode);
        Node middleNext = middle.next;

        middle.next = null;

        Node left = mergeSort(headNode, comparator);
        Node right = mergeSort(middleNext, comparator);

        return merge(left, right, comparator);
    }

    private Node merge(Node firstNode, Node secondNode, Comparator comparator) {
        Node merged = new Node(null);
        Node temp = merged;
        Node tail = head.prev;
        while (firstNode != null && secondNode != null) {
            if (comparator.compare(firstNode.data, secondNode.data) < 0) {
                temp.next = firstNode;
                firstNode.prev = temp;
                firstNode = firstNode.next;
            } else {
                temp.next = secondNode;
                secondNode.prev = temp;
                secondNode = secondNode.next;
            }
            temp = temp.next;
        }
        while (firstNode != null) {
            temp.next = firstNode;
            firstNode.prev = temp;
            firstNode = firstNode.next;
            temp = temp.next;
        }
        while (secondNode != null) {
            temp.next = secondNode;
            secondNode.prev = temp;
            secondNode = secondNode.next;
            temp = temp.next;
            tail = temp;
        }
        return merged.next;
    }

    private Node getMidNode(Node node) {
        Node previousNode = node;
        Node currentNode = node;
        while (currentNode.next != null && currentNode.next.next != null) {
            previousNode = previousNode.next;
            currentNode = currentNode.next.next;
        }
        return previousNode;
    }

    public void printList() {
        Node tmp = head;
        for (int i = 0; i < length; i++) {
            System.out.print(i + ") ");
            System.out.println(tmp.data);
            tmp = tmp.next;
        }
    }

    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        Node tmp = head;
        for (int i = 0; i < length; i++) {
            str.append(tmp.data).append("\n");
            tmp = tmp.next;
        }
        return str.toString();

    }

    public void clearList() {
        head = null;
        length = 0;
    }

    public void save(UserType userType, String fileName) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileName))) {
            writer.write(userType.typeName() + "\n");
            this.forEach(data -> writer.write(data.toString() + "\n"));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void load(UserType userType, String fileName) {
        clearList();
        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String line;
            line = br.readLine();
            if (!userType.typeName().equals(line)) {
                throw new Exception("Wrong file structure");
            }

            while ((line = br.readLine()) != null) {
                add(userType.parseValue(line));
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

}
